alvr/sockets/src/lib.rs

```diff
@@ -93,3 +96,180 @@ fn set_dscp(socket: &Socket, dscp: Option<DscpTos>) {
+    pub recv_buffer_bytes: SocketBufferSize,
+    pub max_packet_size: usize,
+    pub dscp: Option<DscpTos>,
+}
+
+pub enum ServerConnectionResult {
+    Connected(SocketConnection),
+    Restarting,
+}
+
+pub struct SocketConnection {
+    control_socket: ProtoControlSocket,
+    stream_socket: StreamSocket,
+}
+
+impl SocketConnection {
+    // Must be called by the server
+    // Note: the timeout resets after each internal operation
+    pub fn from_client_connection(
+        mut control_socket: ProtoControlSocket,
+        timeout: Duration,
+        stream_config_packet: impl Serialize,
+        socket_config: StreamSocketConfig,
+    ) -> ConResult<Self> {
+        let client_ip = control_socket.inner.peer_addr().to_con()?.ip();
+
+        control_socket.send(&stream_config_packet).to_con()?;
+
+        control_socket
+            .send(&ServerControlPacket::StartStream)
+            .to_con()?;
+
+        let signal = control_socket.recv(timeout)?;
+        if !matches!(signal, ClientControlPacket::StreamReady) {
+            con_bail!("Got unexpected packet waiting for stream ack");
+        }
+
+        let stream_socket = StreamSocketBuilder::connect_to_client(
+            timeout,
+            client_ip,
+            socket_config.port,
+            socket_config.protocol,
+            socket_config.dscp,
+            socket_config.send_buffer_bytes,
+            socket_config.recv_buffer_bytes,
+            socket_config.max_packet_size,
+        )?;
+
+        Ok(Self {
+            control_socket,
+            stream_socket,
+        })
+    }
+
+    // Must be called by the client
+    pub fn from_server_connection(
+        mut control_socket: ProtoControlSocket,
+        timeout: Duration,
+        socket_config: StreamSocketConfig,
+    ) -> ConResult<ServerConnectionResult> {
+        let server_ip = control_socket.inner.peer_addr().to_con()?.ip();
+
+        match control_socket.recv(timeout)? {
+            ServerControlPacket::StartStream => (),
+            ServerControlPacket::Restarting => return Ok(ServerConnectionResult::Restarting),
+            _ => con_bail!("Got unexpected packet waiting for stream start"),
+        }
+
+        let stream_socket_builder = StreamSocketBuilder::listen_for_server(
+            timeout,
+            socket_config.port,
+            socket_config.protocol,
+            socket_config.dscp,
+            socket_config.send_buffer_bytes,
+            socket_config.recv_buffer_bytes,
+        )
+        .to_con()?;
+
+        control_socket
+            .send(&ClientControlPacket::StreamReady)
+            .to_con()?;
+
+        let stream_socket = stream_socket_builder.accept_from_server(
+            server_ip,
+            socket_config.port,
+            socket_config.max_packet_size,
+            timeout,
+        )?;
+
+        Ok(ServerConnectionResult::Connected(Self {
+            control_socket,
+            stream_socket,
+        }))
+    }
```

Comment:
Honestly not sure if this is a good idea if you have to annotate it with 'must be called by x', but idk maybe it's fine

Reply:
To be honest, the comment is unnecessary. because simply by calling one or the other function, the peer assumes the role of client or server, without redundancy